lib/wallaroo/core/sink/connector_sink/connector_sink.pony

```diff
@@ -363,8 +576,130 @@ actor ConnectorSink is Sink
     end
     match barrier_token
     | let sbt: CheckpointBarrierToken =>
-      checkpoint_state(sbt.id)
+      if not _notify.twopc_intro_done then
+        // This sink applies Pony runtime backpressure when disconnected,
+        // so it's quite unlikely that we will get here: sending
+        // messages to us will block the senders.  However, the nature
+        // of backpressure may change over time as the runtime's
+        // backpressure system changes.
+        ifdef "checkpoint_trace" then
+          @printf[I32]("2PC: preemptive abort: connector sink not fully connected\n".cstring())
+        end
+        _abort_decision("connector sink not fully connected",
+          _twopc_txn_id, _twopc_barrier_token)
+
+        _twopc_txn_id = "preemptive txn abort"
+        _twopc_barrier_token = sbt
+        return
+      end
+
+      if _twopc_state is cp.TwoPCFsmStart then
+        let txn_id = _make_txn_id_string(sbt.id)
+        _twopc_txn_id = txn_id
+        // Correctness of the barrier protocol requires us to write
+        // checkpoint state here, not in barrier_fully_acked()
+        checkpoint_state(sbt.id)
+
+        if (_twopc_current_offset > 0) and
+           (_twopc_current_offset == _twopc_last_offset)
+        then
+          // If no data has been processed by the sink since the last
+          // checkpoint, then don't bother with 2PC.
+
+          ifdef "checkpoint_trace" then
+            @printf[I32]("2PC: no data written during this checkpoint interval, skipping 2PC round\n".cstring())
+          end
+          _barrier_coordinator.ack_barrier(this, sbt)
+          _twopc_txn_id = ""
+          _twopc_phase1_commit = true
+          _twopc_state = cp.TwoPCFsm2Commit
+          return
+        end
+
+        let where_list: cp.WhereList =
+          [(1, _twopc_last_offset.u64(), _twopc_current_offset.u64())]
+        let b = cp.TwoPCEncode.phase1(txn_id, where_list)
```

Comment:
Maybe `bs`?  I assume this stands for `bytes`

Reply:
ok